apps/crn-server/src/data-providers/contentful/calendars.data-provider.ts

```diff
@@ -0,0 +1,130 @@
+import {
+  CalendarCreateDataObject,
+  CalendarDataObject,
+  CalendarDataProvider,
+  CalendarUpdateDataObject,
+  FetchCalendarProviderOptions,
+  ListCalendarDataObject,
+  isGoogleLegacyCalendarColor,
+} from '@asap-hub/model';
+import {
+  addLocaleToFields,
+  GraphQLClient,
+  FETCH_CALENDARS,
+  FETCH_CALENDAR_BY_ID,
+  FetchCalendarByIdQuery,
+  FetchCalendarByIdQueryVariables,
+  FetchCalendarsQuery,
+  FetchCalendarsQueryVariables,
+  Environment,
+  CalendarsOrder,
+} from '@asap-hub/contentful';
+
+export type CalendarItem = NonNullable<
+  NonNullable<FetchCalendarsQuery['calendarsCollection']>['items'][number]
+>;
+
+export class CalendarContentfulDataProvider implements CalendarDataProvider {
+  constructor(
+    private contentfulClient: GraphQLClient,
+    private getRestClient: () => Promise<Environment>,
+  ) {}
+
+  async fetchById(id: string): Promise<CalendarDataObject | null> {
+    const { calendars } = await this.contentfulClient.request<
+      FetchCalendarByIdQuery,
+      FetchCalendarByIdQueryVariables
+    >(FETCH_CALENDAR_BY_ID, { id });
+
+    if (!calendars) {
+      return null;
+    }
+
+    return parseGraphQlCalendarToDataObject(calendars);
+  }
+
+  async fetch(
+    options?: FetchCalendarProviderOptions,
+  ): Promise<ListCalendarDataObject> {
+    const { maxExpiration, resourceId } = options || {};
+
+    const { calendarsCollection } = await this.contentfulClient.request<
+      FetchCalendarsQuery,
+      FetchCalendarsQueryVariables
+    >(FETCH_CALENDARS, {
+      limit: 50,
+      skip: 0,
+      order: [CalendarsOrder.NameAsc],
+      where: {
+        ...(maxExpiration ? { expirationDate_lt: maxExpiration } : {}),
+        ...(resourceId ? { resourceId } : {}),
+      },
+    });
+
+    if (!calendarsCollection?.items) {
+      return {
+        total: 0,
+        items: [],
+      };
+    }
+
+    return {
+      total: calendarsCollection.total,
+      items: calendarsCollection.items
+        .filter((x): x is CalendarItem => x !== null)
+        .map(parseGraphQlCalendarToDataObject),
+    };
+  }
+
+  async create(create: CalendarCreateDataObject): Promise<string> {
+    const environment = await this.getRestClient();
+
+    const calendarEntry = await environment.createEntry('calendars', {
+      fields: addLocaleToFields(create),
+    });
+
+    await calendarEntry.publish();
+
+    return calendarEntry.sys.id;
+  }
+
+  async update(id: string, update: CalendarUpdateDataObject): Promise<void> {
+    const environment = await this.getRestClient();
+
+    const calendar = await environment.getEntry(id);
+
+    Object.entries(update).forEach(([fieldName, fieldValue]) => {
```

Comment:
it's probably something we do all over the place, maybe there should be a helper function for that?

Reply:
committed here [moving update function to utils](https://github.com/yldio/asap-hub/pull/2898/commits/6d7e2795673f8de4eac34215a08dd37e6b6828fc)